Content.Client/Audio/AmbientSoundSystem.cs

```diff
@@ -127,56 +113,122 @@ private void ClearSounds()
             _playingSounds.Clear();
         }
 
-        /// <summary>
-        /// Get a list of ambient components in range and determine which ones to start playing.
-        /// </summary>
-        private void SampleNearby(EntityCoordinates coordinates)
+        private (Dictionary<string, List<AmbientSoundComponent>>, HashSet<EntityUid>) GetNearbySources(EntityCoordinates coordinates)
         {
-            var compsInRange = new List<AmbientSoundComponent>();
+            var sourceDict = new Dictionary<string, List<AmbientSoundComponent>>();
+            var set = new HashSet<EntityUid>();
 
-            foreach (var entity in _lookup.GetEntitiesInRange(coordinates, _maxAmbientRange,
-                LookupFlags.Approximate | LookupFlags.IncludeAnchored))
+            foreach (var entity in _lookup.GetEntitiesInRange(coordinates, _maxAmbientRange + RangeBuffer, LookupFlags.IncludeAnchored))
             {
                 if (!EntityManager.TryGetComponent(entity, out AmbientSoundComponent? ambientComp) ||
-                    _playingSounds.ContainsKey(ambientComp) ||
                     !ambientComp.Enabled ||
-                    // We'll also do this crude distance check because it's what we're doing in the active loop above.
                     !EntityManager.GetComponent<TransformComponent>(entity).Coordinates.TryDistance(EntityManager, coordinates, out var range) ||
-                    range > ambientComp.Range - RangeBuffer)
+                    range > ambientComp.Range)
                 {
                     continue;
                 }
 
-                compsInRange.Add(ambientComp);
+                var key = ambientComp.Sound.GetSound();
+
+                if (!sourceDict.ContainsKey(key))
+                    sourceDict[key] = new List<AmbientSoundComponent>(_maxSingleSound);
+
+                sourceDict[key].Add(ambientComp);
+                set.Add(ambientComp.Owner);
             }
 
-            while (_playingSounds.Count < _maxAmbientCount)
+            foreach (var (key, val) in sourceDict)
             {
-                if (compsInRange.Count == 0) break;
+                sourceDict[key] = val.OrderByDescending(x =>
+                    Transform(x.Owner).Coordinates.TryDistance(EntityManager, coordinates, out var dist) ? dist : float.MaxValue).ToList();
+            }
 
-                var comp = _random.PickAndTake(compsInRange);
-                var sound = comp.Sound.GetSound();
+            return (sourceDict, set);
+        }
 
-                if (PlayingCount(sound) >= _maxSingleSound) continue;
+        /// <summary>
+        /// Get a list of ambient components in range and determine which ones to start playing.
+        /// </summary>
+        private void SampleNearby(EntityCoordinates coordinates)
+        {
+            var (compsInRange, compSet) = GetNearbySources(coordinates);
```

Comment:
this can probably be named better considering it doesnt just sample the sounds but actually plays them

Reply:
> this can probably be named better considering it doesnt just sample the sounds but actually plays them

ye ofc